File: python/Adapter/duck.py
from abc import ABC, abstractmethod


class Duck(ABC):

    @abstractmethod
    def quack(self):
        pass

    @abstractmethod
    def fly(self):
        pass


class MallardDuck(Duck):
    def __init__(self):
        pass

    def quack(self):
        print("Quack")

    def fly(self):
        print("I'm flying")

File: c/Adapter/duck.c
#include "duck.h"
void MalDuckNew(MallardDuck *self)
{
    self->quack = MalDuckQuack;
    self->fly = MalDuckFly;
}
void MalDuckQuack(MallardDuck *self)
{
    printf("Quack\n");
}
void MalDuckFly(MallardDuck *self)
{
    printf("I'm flying\n");
}